add LOCATION_UNKNOWN_FILE_ACTION

jobisoft opened the pull request:
I do not know if this is diametral to your design concept, but I would like to have a generic location, so I do not need to add a hard-coded location for each "simple" window iframe. We could later add more options to tweak the general location.

jobisoft committed:
add a generic location

jobisoft committed:
revert not needed change

jobisoft commented:
Hm, I need to check a few things. I close this for now.

rsjtdrjgfuzkfg commented:
This is a very good idea while it is an experiment, but it needs a clear 'legacy' / 'deprecated' comment imho (as WebExtensions using it would depend on implementation details of Thunderbird).

Regarding the code: personally, I'd not name it `generic`, as that sounds like an arbitrary generic location ("Place my UI somewhere, I don't care where"). I'd prefer `raw` or `legacy`. It would also feel nicer to have an object with the same name in the options structure, containing the url and the id, as we can then give a global comment to the whole thing (clarifying that the object is required for the raw/legacy location and changes to the option after adding the frame will not have any effect).

Unrelated note: it would be nice if you could include 'customui' in your commit message for consistency, as this repository contains more than one experiment.

jobisoft committed:
make legacy location configurable and add dedicated UNKNOWN_DOWNLOAD_LOCATION

jobisoft commented:
I updated the PR and renamed the location LEGACY. If you use "squash and merge", my commit messages will be removed and only the message you use during the merge will be used, so you can make sure to have it include "customUI".

I added a dedicated location UNKNOWN_DOWNLOAD_LOCATION which was my main use case to help a new add-on submitted by Opto/Klaus to not use the WindowListener API. I started off with the generic location but the longer I was working on that, the more I came to believe it is the wrong approach for the endusers/developers. A generic location will not tell us, what developers need.

Nevertheless, I think the generic location is useful for developing new locations. I leave it up to you. The generic location is configured via a config option object:

```
messenger.ex_customui.add(
    messenger.ex_customui.LOCATION_LEGACY,
    "download/download.html",
    {
      height: 150,
      hidden: true, //default to hidden and only show if the content-type is an image
      config: {
        windowUrl: "chrome://mozapps/content/downloads/unknownContentType.xhtml",
        glueScript: "download/download-legacy-glue.js",
        referenceElement: "container",
      }
    }
  );
```

The `windowUrl` is the URL we want to inject into. The `glueScript` is a script executed in the host window to collect information we might need as part of the context inside the iFrame. The `referenceElement` is the element after which the iframe should be added in the host window. It can also be returned by the glueScript, if the position might be variable. 

This was the glue script for the UNKNOW_DOWNLOAD_LOCATION before I added it directly: 

```
// The customUI API includes a few pre-defined locations where all the ugly
// core things are taken care of internally and the developer just has to
// provide the WebExtension page which schould be loaded into the provided
// iFrame. If those pre-defined locations are not sufficient, developers can
// use the LOCATION_LEGACY location. This location needs an additional glue
// script to return all the information from the privileged host window, which
// could be needed by the WebExtension page inside the iframe. It must have
// an init() function, which could be async.

async function init(window) {
  //get url from the dialog
  let url = window.dialog.mLauncher.source.spec;
  //get just the query, for example "part=1.2&type=image/jpeg&filename=IMG_0101.jpg"
  let query = window.dialog.mLauncher.source.query;
  
  let data = {};
  for (const part of query.split("&")) {
    const [key,value] = part.split("=");
    data[key] = decodeURIComponent(value);
  }
  data.url = url;
  return data;
}
```

jobisoft committed:
Update parent.js

jobisoft committed:
Update parent.js

jobisoft committed:
rename location

jobisoft committed:
remove legacy location again

jobisoft committed:
Update parent.js

jobisoft committed:
Update parent.js

jobisoft committed:
Update parent.js

jobisoft committed:
Update parent.js
